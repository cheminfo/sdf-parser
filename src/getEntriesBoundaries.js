export function getEntriesBoundaries(string, substring, eol) {
  const res = [];
  let previous = 0;
  let next = 0;
  while (next !== -1) {
    next = string.indexOf(substring, previous);
    if (next !== -1) {
      res.push([previous, next]);
      previous = next =
        string.indexOf(eol, next + substring.length) + eol.length;
    } else {
      res.push([previous, string.length]);
    }
  }
  return res;
}
