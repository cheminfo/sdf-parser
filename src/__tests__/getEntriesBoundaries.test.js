import fs from 'fs';

import { getEntriesBoundaries } from '../getEntriesBoundaries';

let sdf0 = fs.readFileSync(`${__dirname}/test.sdf`, 'utf-8');
let sdf1 = fs.readFileSync(`${__dirname}/test1.sdf`, 'utf-8');
let sdf2 = fs.readFileSync(`${__dirname}/test2.sdf`, 'utf-8');

[sdf0, sdf1, sdf2].forEach((sdf) => {
  let eol = '\n';
  let header = sdf.substr(0, 1000);
  if (header.indexOf('\r\n') > -1) {
    eol = '\r\n';
  } else if (header.indexOf('\r') > -1) {
    eol = '\r';
  }

  test('Split should match regex behavior', () => {
    let sdfParts = sdf.split(new RegExp(`${eol}\\$\\$\\$\\$.*${eol}`));
    expect(sdfParts).toStrictEqual(
      getEntriesBoundaries(sdf, `${eol}$$$$`, eol).map((v) =>
        sdf.substring(...v),
      ),
    );
  });
});
