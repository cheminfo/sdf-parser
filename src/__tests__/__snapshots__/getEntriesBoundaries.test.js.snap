// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should parse sdf files without EOL in the EOF 1`] = `
[
  [
    0,
    1473,
  ],
]
`;
